jupyter_client/manager.py

```diff
@@ -488,6 +510,215 @@ def is_alive(self):
+        """Restarts a kernel with the arguments that were used to launch it.
+
+        Parameters
+        ----------
+        now : bool, optional
+            If True, the kernel is forcefully restarted *immediately*, without
+            having a chance to do any cleanup action.  Otherwise the kernel is
+            given 1s to clean up before a forceful restart is issued.
+
+            In all cases the kernel is restarted, the only difference is whether
+            it is given a chance to perform a clean shutdown or not.
+
+        newports : bool, optional
+            If the old kernel was launched with random ports, this flag decides
+            whether the same ports and connection file will be used again.
+            If False, the same ports and connection file are used. This is
+            the default. If True, new random port numbers are chosen and a
+            new connection file is written. It is still possible that the newly
+            chosen random port numbers happen to be the same as the old ones.
+
+        `**kw` : optional
+            Any options specified here will overwrite those used to launch the
+            kernel.
+        """
+        if self._launch_args is None:
+            raise RuntimeError("Cannot restart the kernel. "
+                               "No previous call to 'start_kernel'.")
+        else:
+            # Stop currently running kernel.
+            await self.shutdown_kernel(now=now, restart=True)
+
+            if newports:
+                self.cleanup_random_ports()
+
+            # Start new kernel.
+            self._launch_args.update(kw)
+            await self.start_kernel(**self._launch_args)
+        return None
+
+    async def _kill_kernel(self):
+        """Kill the running kernel.
+
+        This is a private method, callers should use shutdown_kernel(now=True).
+        """
+        if self.has_kernel:
+
+            # Signal the kernel to terminate (sends SIGKILL on Unix and calls
+            # TerminateProcess() on Win32).
+            try:
+                if hasattr(signal, 'SIGKILL'):
+                    await self.signal_kernel(signal.SIGKILL)
+                else:
+                    await self.kernel.kill()
+            except OSError as e:
+                # In Windows, we will get an Access Denied error if the process
+                # has already terminated. Ignore it.
+                if sys.platform == 'win32':
+                    if e.winerror != 5:
+                        raise
+                # On Unix, we may get an ESRCH error if the process has already
+                # terminated. Ignore it.
+                else:
+                    from errno import ESRCH
+                    if e.errno != ESRCH:
+                        raise
+
+            # Wait until the kernel terminates.
+            try:
+                await asyncio.wait_for(gen.maybe_future(self.kernel.wait()), timeout=5.0)
```

Comment:
Yes - good point.  I think `kernel_async_wait()` would look something like this: https://github.com/takluyver/jupyter_kernel_mgmt/blob/master/jupyter_kernel_mgmt/subproc/manager.py#L56-L66

Does that look right?

Reply:
That sure does look a lot like `finish_shutdown()` so I'll look into refactoring these areas a bit.

Reply:
> Yes - good point. I think `kernel_async_wait()` would look something like this: https://github.com/takluyver/jupyter_kernel_mgmt/blob/master/jupyter_kernel_mgmt/subproc/manager.py#L56-L66
> 
> Does that look right?

Indeed. They even use `asyncio.subprocess.Process`, which would be ideal, but I'm not sure we can have it in `jupyter_client`.

Reply:
Right - the hierarchy of things gets in the way for that.

Btw, I'm finding I need to indicate that the _async_wait() function is done...
```python
    async def _async_wait(self, pollinterval=0.1):
        # Use busy loop at 100ms intervals, polling until the process is
        # not alive.  If we find the process is no longer alive, complete
        # its cleanup via the blocking wait().  Callers are responsible for
        # issuing calls to wait() using a timeout (see _kill_kernel()).
        while self.is_alive():
            await asyncio.sleep(pollinterval)

        done = asyncio.Future()
        done.set_result(None)
        return done
```
otherwise I get failures stemming from its introduction since `asyncio.wait_for()` requires a future.  This seems correct to me, but I honestly lack experience in this area.

Reply:
`asyncio.wait_for` accepts an awaitable - a coroutine, a Task or a Future. `_async_wait` doesn't have to return a result:
```python
    async def _async_wait(self, pollinterval=0.1):
        while self.is_alive():
            await asyncio.sleep(pollinterval)
```
But you need to catch the timeout:
```python
try:
    await asyncio.wait_for(self._async_wait(), timeout=5.0)
except asyncio.TimeoutError:
    # Wait timed out, just log warning but continue - not much more we can do.
    self.log.warning("Wait for final termination of kernel timed out - continuing...")
    pass
```
That is what your code was doing already. I realize my comment above was wrong, there is no `await` in the `wait_for` call.